app_check/integration_test/src/integration_test.cc

```diff
@@ -821,4 +919,191 @@ TEST_F(FirebaseAppCheckTest, TestFunctionsFailure) {
                     firebase::functions::kErrorUnauthenticated);
 }
 
+TEST_F(FirebaseAppCheckTest, TestFirestoreSetGet) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("foo")},
+          {"int", firebase::firestore::FieldValue::Integer(123)}}),
+      "document.Set");
+  firebase::Future<firebase::firestore::DocumentSnapshot> future =
+      document.Get(firebase::firestore::Source::kServer);
+  WaitForCompletion(future, "document.Get");
+  EXPECT_NE(future.result(), nullptr);
+  EXPECT_THAT(future.result()->GetData(),
+              UnorderedElementsAre(
+                  Pair("str", firebase::firestore::FieldValue::String("foo")),
+                  Pair("int", firebase::firestore::FieldValue::Integer(123))));
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreSetGetFailure) {
+  // Don't set up AppCheck
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  // Both operations should fail because AppCheck isn't configured.
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("badfoo")},
+          {"int", firebase::firestore::FieldValue::Integer(456)}}),
+      "document.Set", firebase::firestore::kErrorPermissionDenied);
+  WaitForCompletion(document.Get(firebase::firestore::Source::kServer),
+                    "document.Get",
+                    firebase::firestore::kErrorPermissionDenied);
+
+  CleanupFirestore(firebase::firestore::kErrorPermissionDenied);
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreTransaction) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("start")}}),
+      "document.Set 0");
+
+  std::vector<firebase::firestore::MapFieldValue> document_snapshots;
+  firebase::firestore::ListenerRegistration registration =
+      document.AddSnapshotListener(
+          [&](const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            EXPECT_EQ(error_code, firebase::firestore::kErrorOk);
```

Comment:
Consider adding a `SCOPED_TRACE()` here with a message that includes the size of the vector. That way, if one of the expectations fails it is clear which invocation(s) suffered from the failure.

Reply:
Done